nucypher-core/src/dkg.rs

```diff
@@ -21,6 +109,280 @@ pub enum FerveoVariant {
+    }
+
+    impl AsRef<[u8]> for RequestSharedSecret {
+        /// View this shared secret key as a byte array.
+        #[inline]
+        fn as_ref(&self) -> &[u8] {
+            self.0.as_bytes()
+        }
+    }
+
+    impl fmt::Display for RequestSharedSecret {
+        /// Format shared secret information.
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "RequestSharedSecret...")
+        }
+    }
+
+    /// A request public key.
+    #[derive(PartialEq, Eq, Hash, Copy, Clone, Debug, Serialize, Deserialize)]
+    pub struct RequestPublicKey(pub(crate) PublicKey);
+
+    /// Implementation of request public key
+    impl RequestPublicKey {
+        /// Convert this public key to a byte array.
+        #[inline]
+        pub fn to_bytes(&self) -> [u8; 32] {
+            self.0.to_bytes()
+        }
+
+        /// View this public key as a byte array.
+        #[inline]
+        pub fn as_bytes(&self) -> &[u8; 32] {
+            self.0.as_bytes()
+        }
+    }
+
+    impl AsRef<[u8]> for RequestPublicKey {
+        /// View this public key as a byte array.
+        #[inline]
+        fn as_ref(&self) -> &[u8] {
+            self.0.as_bytes()
+        }
+    }
+
+    impl fmt::Display for RequestPublicKey {
+        /// Format public key information.
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            let bytes = self.as_ref();
+            let bytes = if bytes.len() > 8 { &bytes[..8] } else { bytes };
+            write!(f, "RequestPublicKey: {}", hex::encode(bytes),)
+        }
+    }
+
+    impl<'a> From<&'a RequestSecretKey> for RequestPublicKey {
+        /// Compute corresponding [`RequestPublicKey`].
+        fn from(secret: &'a RequestSecretKey) -> RequestPublicKey {
+            let public_key = PublicKey::from(&secret.0);
+            RequestPublicKey(public_key)
+        }
+    }
+
+    impl<'a> ProtocolObjectInner<'a> for RequestPublicKey {
+        fn version() -> (u16, u16) {
+            (1, 0)
+        }
+
+        fn brand() -> [u8; 4] {
+            *b"TRPk"
+        }
+
+        fn unversioned_to_bytes(&self) -> Box<[u8]> {
+            messagepack_serialize(&self)
+        }
+
+        fn unversioned_from_bytes(
+            minor_version: u16,
+            bytes: &[u8],
+        ) -> Option<Result<Self, String>> {
+            if minor_version == 0 {
+                Some(messagepack_deserialize(bytes))
+            } else {
+                None
+            }
+        }
+    }
+
+    impl<'a> ProtocolObject<'a> for RequestPublicKey {}
+
+    /// A request secret key.
+    #[derive(ZeroizeOnDrop)]
+    pub struct RequestSecretKey(pub(crate) StaticSecret);
+
+    impl RequestSecretKey {
+        /// Perform diffie-hellman
+        pub fn diffie_hellman(&self, their_public_key: &RequestPublicKey) -> RequestSharedSecret {
+            let shared_secret = self.0.diffie_hellman(&their_public_key.0);
+            RequestSharedSecret(shared_secret)
```

Comment:
As a defense-in-depth measure, it's recommended to hash/HKDF the DH result so we get rid of any potential structure inherent to a DH result (which after all is a simple EC point). There are several way to do this, see e.g. https://github.com/nucypher/rust-umbral/blob/975b0f395b8a09e9b5cdc7fe6ba6f25ff171cd78/umbral-pre/src/hashing_ds.rs#L30

Also related to this, perhaps a better API here is changing the method name to `derive_shared_secret` or `shared_secret_with`, so we don't get lost on low-level details (like the DH exchange) and focus on what we're actually creating here, which is a shared secret bytestring.

Reply:
If I understand it correctly, we're using the direct bytes representation of the DH shared secret class from the dalek crate, which I believe it's just a curve point (specifically the x coordinate). While this could be enough as a key, we can't be sure that there's a slight weakness stemming from the not 100% uniform output distribution (e.g. there are some x coordinates with exactly 0 probability of occurrence). The solution for this is to pass the output through a KDF. This works as long as the entropy bits you are passing as input are roughly the same (or more, of course) as those you expect later as a symmetric key, which is the case here.

A much better explanation here:
https://crypto.stackexchange.com/questions/10660/why-must-curve25519-shared-secret-be-hashed

Reply:
✔️ Updated - let me know if I understood that correctly.